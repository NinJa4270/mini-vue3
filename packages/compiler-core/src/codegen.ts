import { isString } from "@ninja/shared";
import { CompoundExpressionNode, ElementNode, InterpolationNode, JSChildNode, NodeTypes, PlainElementNode, RootNode, SimpleExpressionNode, TemplateChildNode, TextNode } from "./ast";
import { CREATE_ELEMENT_VNODE, helperNameMap, TO_DISPLAY_STRING } from "./runtimeHelpers";

interface CodegenContext {
    code: string
    helper(key: symbol): string
    push(code: string): void
}
type CodegenNode = TemplateChildNode | JSChildNode

const aliasHelper = (s: symbol) => `${helperNameMap[s]}: _${helperNameMap[s]}`

export function generate(ast: RootNode) {
    const context = createCodegenContext(ast)
    const { push } = context
    push(`\n`)

    genFunctionPreamble(ast, context)
    const functionName = "render"
    const args = ['_ctx', '_cache']
    const signature = args.join(', ')
    push(` function ${functionName}(${signature}) {`)
    push(`return `)

    if (ast.codegenNode) {
        genNode(ast.codegenNode, context)
    } else {
        push(`null`)
    }
    push(`}`)
    return {
        code: context.code
    }
}

function genFunctionPreamble(ast: RootNode, context: CodegenContext) {
    const VueBinging = `Vue`
    const { push } = context
    if (ast.helpers.length > 0) {
        push(
            `const { ${ast.helpers.map(aliasHelper).join(", ")}} = ${VueBinging} `
        );
    }
    push(`\n`)
    push(`return`)
}


function genNode(node: CodegenNode, context: CodegenContext) {
    if (!node) return
    switch (node.type) {
        case NodeTypes.TEXT:
            genText(node as TextNode, context)
            break
        case NodeTypes.ELEMENT:
            genElement(node as PlainElementNode, context)
            break
        case NodeTypes.SIMPLE_EXPRESSION:
            genExpression(node as SimpleExpressionNode, context)
            break
        case NodeTypes.INTERPOLATION:
            genInterpolation(node as InterpolationNode, context)
            break
        case NodeTypes.COMPOUND_EXPRESSION:
            genCompoundExpression(node as CompoundExpressionNode, context)
            break
        default:
            break
    }
}

// 元素类型
function genElement(node: ElementNode, context: CodegenContext) {
    const { push, helper } = context
    const { tag, children, props } = node
    push(` ${helper(CREATE_ELEMENT_VNODE)}(`)
    genNodeList(genNullable([tag, props, children]), context)
    // genNode(children as unknown as CodegenNode, context)
    push(`)`)
}

function genNodeList(nodes: unknown[], context: CodegenContext) {
    const { push } = context
    for (let i = 0; i < nodes.length; i++) {
        const node = nodes[i];
        if (isString(node)) {
            push(`${node}`)
        } else {
            genNode(node as CodegenNode, context)
        }

        if (i < nodes.length - 1) {
            push(", ")
        }
    }
}

function genNullable(args: unknown[]) {
    return args.map(arg => arg || "null")
}

// 文字类型
function genText(node: TextNode, context: CodegenContext) {
    const { push } = context
    push(` "${node.content}"`)
}

// 插值类型
function genInterpolation(node: InterpolationNode, context: CodegenContext) {
    const { push, helper } = context
    push(` ${helper(TO_DISPLAY_STRING)}(`)
    genNode(node.content, context)
    push(`)`)
}

// 插值内表达式类型
function genExpression(node: SimpleExpressionNode, context: CodegenContext) {
    const { content } = node
    const { push } = context
    push(content)
}
// 复合类型
function genCompoundExpression(node: CompoundExpressionNode, context: CodegenContext) {
    const { push } = context
    const { children } = node
    for (let i = 0; i < children.length; i++) {
        const child = children[i]
        if (isString(child)) {
            push(child)
        } else {
            genNode(child, context)
        }
    }
}

function createCodegenContext(ast: RootNode): CodegenContext {
    const context: CodegenContext = {
        code: ``,
        push(code) {
            context.code += code
        },
        helper(key) {
            return `_${helperNameMap[key]}`
        },
    }
    return context
}