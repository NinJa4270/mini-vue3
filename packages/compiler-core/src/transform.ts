import { NodeTypes, RootNode, TemplateChildNode, ParentNode } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";


export type NodeTransform = (node: RootNode | TemplateChildNode, context: TransformContext) => void | (() => void) | (() => void)[]

interface TransformOptions {
    nodeTransforms?: NodeTransform[]
}

export function transform(root: RootNode, options: TransformOptions) {
    const context = createTransformContext(root, options)
    // 遍历 - 深度优先
    traverseNode(root, context)

    createRootCodegen(root)

    root.helpers = [...context.helpers.keys()]
}

function createRootCodegen(root: RootNode) {
    const child = root.children[0]
    if (child.type === NodeTypes.ELEMENT) {
        root.codegenNode = child.codegenNode
    } else {
        root.codegenNode = root.children[0]
    }
}

function traverseNode(node: RootNode | TemplateChildNode, context: TransformContext) {
    const { nodeTransforms } = context
    // 修改值
    const exitFns: any = [] // 退出时 需要执行的插件
    for (let i = 0; i < nodeTransforms.length; i++) {
        const onExit = nodeTransforms[i](node, context)
        if (onExit) exitFns.push(onExit)
    }
    switch (node.type) {
        case NodeTypes.INTERPOLATION:
            context.helper(TO_DISPLAY_STRING)
            break
        case NodeTypes.ROOT:
        case NodeTypes.ELEMENT:
            // context.helper(CREATE_ELEMENT_BLOCK)
            traverseChildren(node, context)
            break
    }

    let i = exitFns.length
    while (i--) {
        exitFns[i]()
    }
}

function traverseChildren(parent: ParentNode, context: TransformContext) {
    const children = parent.children
    if (children) {
        for (let i = 0, len = children.length; i < len; i++) {
            const child = children[i]
            traverseNode(child, context)
        }
    }
}

export interface TransformContext {
    nodeTransforms: NodeTransform[],
    helpers: Map<symbol, number>
    helper<T extends symbol>(name: T): T
}

function createTransformContext(root: RootNode, { nodeTransforms = [] }: TransformOptions): TransformContext {
    const context: TransformContext = {
        nodeTransforms,
        helpers: new Map(),
        helper(name) {
            const count = context.helpers.get(name) || 0
            context.helpers.set(name, count + 1)
            return name
        },
    }
    return context
}