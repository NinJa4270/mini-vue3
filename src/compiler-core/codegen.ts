import { RootNode, TextNode } from "./ast";

interface CodegenContext {
    code: string
    push(code: string): void
}

export function generate(ast: RootNode) {
    const context = createCodegenContext(ast)
    const { push } = context
    push(`return`)
    const functionName = "render"
    const args = ['_ctx', '_cache']
    const signature = args.join(', ')
    push(` function ${functionName}(${signature}) {`)
    push(`return `)
    genCode(ast.codegenNode as TextNode, context)
    push(`}`)
    return {
        code: context.code
    }
}

function genCode(codegenNode: TextNode, context: CodegenContext) {
    context.code += ` "${codegenNode.content}"`
}

function createCodegenContext(ast: RootNode): CodegenContext {
    const context = {
        code: ``,
        push(code: string) {
            context.code += code
        }
    }
    return context
}