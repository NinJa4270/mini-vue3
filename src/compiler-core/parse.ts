import { createRoot, ElementNode, InterpolationNode, NodeTypes, TemplateChildNode } from "./ast";

export interface ParserContext {
    source: string
}

export function baseParse(content: string) {
    const context = createParserContext(content)
    return createRoot(parseChildren(context))
}

function parseChildren(context: ParserContext): TemplateChildNode[] {
    const nodes: TemplateChildNode[] = []
    const s = context.source
    let node: TemplateChildNode | undefined = undefined
    if (s.startsWith("{{")) {
        node = parseInterpolation(context)
    } else if (s[0] === '<') {
        if (/[a-z]/i.test(s[1])) {
            node = parseElement(context)
        }
    }
    if (node) {
        nodes.push(node)
    }
    return nodes
}

// 处理元素
function parseElement(context: ParserContext): ElementNode {
    // <div></div>
    // 解析tag
    const element = parseTag(context, TagType.Start)
    // 删除结束
    parseTag(context, TagType.End)
    
    return element
}


const enum TagType {
    Start,
    End
}
// 解析 tag  开始标签/结束标签
function parseTag(context: ParserContext, type: TagType.Start,): ElementNode
function parseTag(context: ParserContext, type: TagType.End): void
function parseTag(context: ParserContext, type: TagType): ElementNode | undefined {
    const match: any = /^<\/?([a-z]*)/i.exec(context.source) // 结束
    const tag = match[1]
    // 删除解析结果
    advanceBy(context, match[0].length) // <div
    advanceBy(context, 1) // >

    if (type === TagType.End) {
        return
    }
    return {
        type: NodeTypes.ELEMENT,
        tag
    }
}

// 处理插值
function parseInterpolation(context: ParserContext): InterpolationNode {
    // {{  message}}
    // delimiters
    const open = "{{"
    const close = "}}"

    const closeIndex = context.source.indexOf(close, open.length)
    advanceBy(context, open.length)

    const rawContentLength = closeIndex - open.length
    const rawContent = context.source.slice(0, rawContentLength)
    const content = rawContent.trim()
    advanceBy(context, rawContentLength)

    advanceBy(context, close.length)

    return {
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content
        }
    }
}

function advanceBy(context: ParserContext, numberOfCharacters: number): void {
    const { source } = context
    context.source = source.slice(numberOfCharacters)
}


function createParserContext(content: string): ParserContext {
    return {
        source: content
    }
}