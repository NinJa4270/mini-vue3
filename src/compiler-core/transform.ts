import { NodeTypes, RootNode, TemplateChildNode, ParentNode } from "./ast";


export type NodeTransform = (node: RootNode | TemplateChildNode) => void | (() => void) | (() => void)[]

interface TransformOptions {
    nodeTransforms?: NodeTransform[]
}


export function transform(root: RootNode, options: TransformOptions) {
    const context = createTransformContext(root, options)
    // 遍历 - 深度优先
    traverseNode(root, context)

    createRootCodegen(root)

}

function createRootCodegen(root: RootNode) {
    root.codegenNode = root.children[0]
}

function traverseNode(node: RootNode | TemplateChildNode, context: TransformContext) {
    const { nodeTransforms } = context
    // 修改值
    for (let i = 0; i < nodeTransforms.length; i++) {
        nodeTransforms[i](node)
    }
    if (node.type === NodeTypes.ROOT || node.type === NodeTypes.ELEMENT) {
        traverseChildren(node, context)
    }
}

function traverseChildren(parent: ParentNode, context: TransformContext) {
    const children = parent.children
    if (children) {
        for (let i = 0, len = children.length; i < len; i++) {
            const child = children[i]
            traverseNode(child, context)
        }
    }
}

interface TransformContext {
    nodeTransforms: NodeTransform[]
}

function createTransformContext(root: RootNode, { nodeTransforms = [], }: TransformOptions): TransformContext {
    const context: TransformContext = {
        nodeTransforms,
    }
    return context
}