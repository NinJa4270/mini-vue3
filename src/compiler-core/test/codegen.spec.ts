import { expect, test, describe } from 'vitest'
import { generate } from '../codegen'
import { baseParse } from '../parse'
import { transform } from '../transform'
describe('codegen', () => {

    test('codegen', () => {
        const ast = baseParse("hi")
        transform(ast, {})
        const { code } = generate(ast)
        // 快照测试
        expect(code).matchSnapshot()
    })
})
