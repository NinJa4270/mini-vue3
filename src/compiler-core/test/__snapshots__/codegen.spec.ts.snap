// Vitest Snapshot v1

exports[`codegen > codegen 1`] = `"return function render(_ctx, _cache) {return  \\"hi\\"}"`;
